### PR TITLE
[TSE] Implement Health Service

### LINKED ISSUE #16: [TSE] Implement Health Service
The task involves implementing a Health Check Service that is responsible for providing the necessary data to assess the health and operational status of TSE (Trading Strategy Execution) instances. This service will ensure that all TSE instances are functioning correctly and provide vital information for monitoring and maintenance.

### To Do
- Endpoint Implementation:
  - Create an API endpoint that returns the health status of TSE instances.

### Deliverables
- [ ] All tasks completed
- [ ] Unit tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by MWBlocky**
## Description

This Pull Request introduces the implementation of a Health Check Service for the TSE (Trading Strategy Execution) instances. This service is designed to provide the necessary data to assess the health and operational status of TSE instances, ensuring that all instances are functioning correctly and supplying vital information for monitoring and maintenance.

## Summary of changes

- **Implemented `HealthService` module**.
- **Created Health API Endpoint**: Added a new endpoint to return the health status of TSE instances.
- **Database Health Check**: Implemented checks to ensure that all required database tables are present and operational.
- **Exchange Health Check**: Implemented checks to verify the configurations and connections to various exchanges used by TSE.
- **Unit tests added**: Verified the correct functionality of the health check methods and the overall service.

## Related issues

Fixes #16

### CODE REVIEW

File: packages/tse-backend/src/modules/health/health.service.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    @InjectEntityManager() private entityManager: EntityManager,
+    private readonly dataSource: DataSource,
+    private readonly configService: ConfigService,
+    private readonly ccxtGateway: CcxtGateway,
+  ) {}
+
+  async geHealthStatuses() {
+    return {
+      db: await this.checkDbHealth(),
+      exchanges: await this.checkExchanges(),
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/health/health.service.ts:23]: Maybe add number of strategies running?

File: packages/tse-backend/src/modules/health/health.service.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      (table) => !existingTables.includes(table),
+    );
+    return { missingTables };
+  }
+
+  async checkExchanges() {
+    this.logger.debug('Checking exchanges configuration health...');
+
+    const expectedExchanges = buildExchangeConfigs(this.configService);
+    const actualExchangesIterator = this.ccxtGateway.getExchangesNames();
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/health/health.service.ts:82]: Extract database functions to a new service

**Review by piotrswierzy — CHANGES_REQUESTED** [review]

File: packages/tse-backend/src/modules/health/health.service.ts (hunk 1 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/health/health.service.ts]: In commit 17cc74a0, I added a TODO annotation. I will implement this service after the pull requests related to the strategies have been merged

**Review by posix4e — CHANGES_REQUESTED** [review]
So it's important that health checks are run passively, automatically and routinely. Not on demand.

So when a user hits a /health endpoint, it should just retreive counters

For instance

    time since last database read
    time since last database change
    time since last change in time series data
    ...

so /health is retrieving this information not causing them to be updated.

**Comment by MWBlocky** [comment]
@posix4e, I'm evaluating some solutions for monitoring database operations in our project. Here are the options:

1. **NestJS Subscribers** ([NestJS Documentation](https://docs.nestjs.com/techniques/database#subscribers)):
   - You can react to `INSERT`, `UPDATE`, and `DELETE` operations by storing relevant data, which could be saved to Redis.
   - There is no built-in support for subscribing to `SELECT` operations. This can be addressed by adding an interceptor and applying it to methods such as `find`.

2. **pg_stat_statements** ([PostgreSQL Documentation](https://www.postgresql.org/docs/current/pgstatstatements.html)):
   - PostgreSQL does not store historical statistics by default. This can be achieved using the `pg_stat_statements` extension, but it requires additional configuration in `postgresql.conf`, which can be challenging during deployment.

3. **Custom Logging**:
   - You could also add a custom logger to each operation, which would not only log to the console but also store data that could be returned by a health check endpoint.


Do any of these solutions look suitable for our implementation? Or do you have other recommendations?

File: packages/tse-backend/src/common/utils/aspect/database/database.cache.util.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+      ctx.methodName.startsWith(prefix)
+    );
+
+    if (strategy) {
+      await this.strategies[strategy].execute(this.cacheManager);
+    } else {
+      this.logger.error(`No cache aspect strategy found for ${ctx.methodName}`);
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (MWBlocky) [reply packages/tse-backend/src/common/utils/aspect/database/database.cache.util.ts]: Fixed: 1661ccc9

**Pull request merged**